monitor/service.go

```diff
@@ -0,0 +1,306 @@
+	"os"
+	"sort"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/influxdb/influxdb/influxql"
+)
+
+// Client is the interface modules must implement if they wish to register with monitor.
+type Client interface {
+	Statistics() (map[string]interface{}, error)
+	Diagnostics() (map[string]interface{}, error)
+}
+
+// Service represents an instance of the monitor service.
+type Service struct {
+	wg            sync.WaitGroup
+	done          chan struct{}
+	mu            sync.Mutex
+	registrations []*clientWithMeta
+
+	hostname  string
+	clusterID uint64
+	nodeID    uint64
+
+	storeEnabled  bool
+	storeDatabase string
+	storeAddress  string
+	storeInterval time.Duration
+
+	expvarAddress string
+
+	Logger *log.Logger
+}
+
+// NewService returns a new instance of the monitor service.
+func NewService(c Config) *Service {
+	return &Service{
+		registrations: make([]*clientWithMeta, 0),
+		storeEnabled:  c.StoreEnabled,
+		storeDatabase: c.StoreDatabase,
+		storeAddress:  c.StoreAddress,
+		storeInterval: time.Duration(c.StoreInterval),
+		expvarAddress: c.ExpvarAddress,
+		Logger:        log.New(os.Stderr, "[monitor] ", log.LstdFlags),
+	}
+}
+
+// Open opens the monitoring service, using the given clusterID, node ID, and hostname
+// for identification purposes.
+func (s *Service) Open(clusterID, nodeID uint64, hostname string) error {
+	s.Logger.Printf("starting monitor service for cluster %d, host %s", clusterID, hostname)
+	s.clusterID = clusterID
+	s.nodeID = nodeID
+	s.hostname = hostname
+
+	// Self-register Go runtime stats.
+	s.Register("runtime", nil, &goRuntime{})
+
+	// If enabled, record stats in a InfluxDB system.
+	if s.storeEnabled {
+		s.Logger.Printf("storing in %s, database '%s', interval %s",
+			s.storeAddress, s.storeDatabase, s.storeInterval)
+
+		// Ensure database exists.
+		values := url.Values{}
+		values.Set("q", fmt.Sprintf("CREATE DATABASE IF NOT EXISTS %s", s.storeDatabase))
+		resp, err := http.Get(s.storeAddress + "/query?" + values.Encode())
+		if err != nil {
+			return fmt.Errorf("failed to create monitoring database on %s: %s", s.storeAddress, err.Error())
+		}
+		if resp.StatusCode != http.StatusOK {
+			return fmt.Errorf("failed to create monitoring database on %s, received code: %d",
+				s.storeAddress, resp.StatusCode)
+		}
+		s.Logger.Printf("succesfully created database %s on %s", s.storeDatabase, s.storeAddress)
+
+		// Start periodic writes to system.
+		s.wg.Add(1)
+		go func() {
+			defer s.wg.Done()
+			tick := time.NewTicker(s.storeInterval)
+			for {
+				select {
+				case <-tick.C:
+					if err := s.storeStatistics(); err != nil {
+						s.Logger.Printf("failed to write statistics to %s: %s", s.storeAddress, err.Error())
+					}
+				case <-s.done:
+					s.Logger.Printf("terminating storage of statistics to %s", s.storeAddress)
+					return
+				}
+
+			}
+		}()
+	}
+
+	// If enabled, expose all expvar data over HTTP.
+	if s.expvarAddress != "" {
+		listener, err := net.Listen("tcp", s.expvarAddress)
```

Comment:
Using the `pprof` endpoint sounds like a good idea, in that case it might just work, and I may not need to create a HTTP service here. Let me check.